### PR TITLE
Synchronise Sticker Pack setting fields

### LINKED ISSUE #13193: Synchronise Sticker Pack setting fields
Implement functionality that handles the syncing of the `stickers/packs-installed`, `stickers/packs-pending` and `stickers/recent-stickers` settings.

### Background

In `status-go` syncing of settings has been implemented, and additional setting sync *should be* a relatively straightforward process. The initial work for syncing of the `stickers/packs-installed`, `stickers/packs-pending` and `stickers/recent-stickers` settings has been implemented on the `status-go` side.

See https://github.com/status-im/status-go/pull/2478 and https://github.com/status-im/status-react/pull/13053

**NOTE:** [Issue 8](https://github.com/status-im/status-react/pull/13053#issuecomment-1065179963) has been resolved, however additional work needs to be done on these setting fields to get robust behaviour:

In my initial work on syncing sticker packs the syncing seemed to work, sometimes the stickers didn't appear, but the space where the stickers should appear was there. I don't know if this was an issue with my implementation or my dev env or what exactly. That'll be your job 🛠️ 

An additional thing to consider is the handling of differences between the sticker pack data payloads.
  - Example: `stickers/packs-installed` may contain 3 sticker packs on device A and 4 sticker packs on device B, and both sets have 2 sticker packs in common and the other sticker packs are unique to their respective devices. What should happen when syncing this data?
  - A potential solution would be to combine the data into a merged set

### `status-go` Implementation

Currently the `stickers/packs-installed`, `stickers/packs-pending` and `stickers/recent-stickers` setting have their sync functionality disabled:

```go
StickersPacksInstalled = SettingField{
	reactFieldName: "stickers/packs-installed",
	dBColumnName:   "stickers_packs_installed",
	valueHandler:   JSONBlobHandler,
	// TODO resolve issue 8 https://github.com/status-im/status-react/pull/13053#issuecomment-1065179963
	syncProtobufFactory: &SyncProtobufFactory{
		inactive:          true, // Remove after issue is resolved
		fromInterface:     stickersPacksInstalledProtobufFactory,
		fromStruct:        stickersPacksInstalledProtobufFactoryStruct,
		valueFromProtobuf: BytesFromSyncProtobuf,
		protobufType:      protobuf.SyncSetting_STICKERS_PACKS_INSTALLED,
	},
}
StickersPacksPending = SettingField{
	reactFieldName: "stickers/packs-pending",
	dBColumnName:   "stickers_packs_pending",
	valueHandler:   JSONBlobHandler,
	// TODO resolve issue 8 https://github.com/status-im/status-react/pull/13053#issuecomment-1065179963
	syncProtobufFactory: &SyncProtobufFactory{
		inactive:          true, // Remove after issue is resolved
		fromInterface:     stickersPacksPendingProtobufFactory,
		fromStruct:        stickersPacksPendingProtobufFactoryStruct,
		valueFromProtobuf: BytesFromSyncProtobuf,
		protobufType:      protobuf.SyncSetting_STICKERS_PACKS_PENDING,
	},
}
StickersRecentStickers = SettingField{
	reactFieldName: "stickers/recent-stickers",
	dBColumnName:   "stickers_recent_stickers",
	valueHandler:   JSONBlobHandler,
	// TODO resolve issue 8 https://github.com/status-im/status-react/pull/13053#issuecomment-1065179963
	syncProtobufFactory: &SyncProtobufFactory{
		inactive:          true, // Remove after issue is resolved
		fromInterface:     stickersRecentStickersProtobufFactory,
		fromStruct:        stickersRecentStickersProtobufFactoryStruct,
		valueFromProtobuf: BytesFromSyncProtobuf,
		protobufType:      protobuf.SyncSetting_STICKERS_RECENT_STICKERS,
	},
}
```

To reactivate the `stickers/packs-installed`, `stickers/packs-pending` and `stickers/recent-stickers` functionality in `status-go` simply remove the `inactive` field from the `SyncProtobufFactory`s

### PR DESCRIPTION
**Pull request by audriu**
Synchronize Sticker Pack setting fields
fixes #13193

### Summary
Implement functionality that handles the syncing of the stickers/recent-stickers settings

#### Platforms
- Android
- iOS

##### Functional
- 1-1 chat
- public chats
- group chats

### Steps to test
Enable synchronization as described in this comment:
https://github.com/status-im/status-react/pull/13053#issuecomment-1065179963

Open Status
Send stickers and see them added to recent stickers and synchronized in another device with the same account.

status: ready

### CODE REVIEW

**Comment by pavloburykh** [comment]
@audriu thanx for the PR. Please, take look at following issue

### ISSUE 1 App crashes after purchasing of stickerpack when screen with stickerpack is opened during completion of transaction


**Prerequisites:** 
For purposes of reproduction of this bug I created user with 20 STT in order you can purchase Stickerpack on Ropsten. 
Recovery seed phrase: van process sugar bike coin damage prison wish tube maximum average reveal

**Steps:** 

1. Recover given above multiaccount
2. Start 1-1 chat
3. Tap sticker icon
4. Tap "Get stickers"
5. Purchase  "Tozemoon" stickerpack
6. Do not close app/stickerpack screen until transaction is completed (spinner is presented)
7. Observe the result after purchase is completed. 

**Actual result:** app crashes. I failed to reproduce this bug in other PRs so I suppose it might be connected to current PR changes. 

**Expected result:** app does not crash. Install button appears after purchase is completed. 

OS: Android/IOS


https://user-images.githubusercontent.com/97245802/171391593-f02c8b52-81b9-4226-a750-9198e719d72a.mp4

[geth.log](https://github.com/status-im/status-react/files/8813526/geth.log)

**Comment by pavloburykh** [comment]
@audriu thanx for the fixes. Please, take a look at Issue 2

### ISSUE 2 Synced data does not appear until relogin in case if there is a bought stickerpack among synced data

**Steps:** 

1. Recover same account on Device 1 and Device 2 
2. On Device 1 buy and install a sticker pack (do not install free sticker pack, as bug is being reproduced only with bought ones)
3. Generate some other data on Device 1 that will be synced to Device 2, for example: add some favourite bookmarks, change wallet currency, add profile picture. 
4. Pair and sync Device 1 and Device 2
5. Pay attention of new data appeared on Device 2

**Actual result**: synced settings and data hasn't appeared on Device 2. It appears only after re-login. 

**Expected result:** synced data should appear on Device 2 right after sync is completed. 

On below video Device 1 (the one with generated data) is on the left side, Device 2 (recipient) is on the right

https://user-images.githubusercontent.com/97245802/172634590-dc55bcb3-2eb9-4752-a2d5-a4a9fd9801ab.mp4

**Comment by pavloburykh** [comment]
### ISSUE 3 Messages are not delivered in group chats

**Steps:** 
1. Login different multiaccounts on Device 1 and Device 1 
2. Add Device 2 to contacts on Device 1
3. Start group chat on Device 1 with Device 2 
4. Send some messages into group chat on Device 1 and Device 2
5. Pay attention if messages are delivered to receiver

**Actual result:** no messages received in group chat

**Expected result:** messages are received

https://user-images.githubusercontent.com/97245802/173349051-3b0b1417-8b6d-4633-880c-96c6dc05c1f2.mp4


### ISSUE 4 Group chat re-appears in chat list on IOS after removing/leaving the chat

**Steps:** 

1. On IOS device start group chat, invite some other user into it
2. Leave the group chat
3. Put app to the background or perform re-login
4. Pay attention at chat list

**Actual result:** deleted chat has re-appeared in chat list. 

**Expected result:** deleted chat does not re-appear


https://user-images.githubusercontent.com/97245802/173349484-f61961a7-4580-4905-91fa-08ef10474476.mp4

**Comment by pavloburykh** [comment]
@audriu Thanx. Issue 3 and Issue 4 fixed after rebase of backend. For Issue 2 I will create a separate Issue in order nor to block merging of this PR.

File: src/status_im/multiaccounts/update/core.cljs
Comment (flexsurfer) [inline src/status_im/multiaccounts/update/core.cljs]: its little bit hard to read, here my suggestion, but it should be tested before merge 
```suggestion
  (mapv #(assoc % :url (-> (get stickers-from-db (:packID %)
                                                          (:stickers)
                                                          (filter (fn [sticker-db] (= (:hash sticker-db) (:hash %))))
                                                          (first)
                                                          (:url))))
                synced-stickers))
```

File: src/status_im/multiaccounts/update/core.cljs
Comment (flexsurfer) [inline src/status_im/multiaccounts/update/core.cljs]: we have types/js-cljs it has keywordize option

File: src/status_im/multiaccounts/update/core.cljs
Comment (audriu) [reply src/status_im/multiaccounts/update/core.cljs]: Thanks for the review. I applied your recommendation (except used ->> because filter). It really looks more readable this way. Also tested - worked the same.

**Review by cammellos — APPROVED** [review]

**Comment by pavloburykh** [comment]
@audriu @cammellos @flexsurfer do we need to re-run e2e and make another round of manual testing due to recent changes?

**Comment by audriu** [comment]
@pavloburykh those changes were just code style. Logic did not change. Although I tested the affected part it still works. Thanks.

**Comment by Samyoul** [comment]
@pavloburykh @audriu do you know if the following scenarios work?

## Scenario 1

- device 2 adds a sticker pack
- device 1 and device 2 pair
- device 1 adds a different sticker pack
- device 2 receives new sticker pack shows in list of sticker packs
- device 2 logs out
- device 2 logs in
- device 2 shows 2 sticker packs (original sticker pack from device 2 and new sticker pack from device 1)

## Scenario 2

- device 2 adds a sticker pack
- device 1 adds a different sticker pack
- device 1 and device 2 pair
- device 1 receives the sticker pack from device 2 shows in list of device 1 sticker packs
- device 2 receives the sticker pack from device 1 shows in list of device 2 sticker packs
- device 1 and 2 log out
- device 1 and 2 log in
- device 1 shows 2 sticker packs (original sticker pack from device 1 and new sticker pack from device 2)
- device 2 shows 2 sticker packs (original sticker pack from device 2 and new sticker pack from device 1)

**Comment by audriu** [comment]
@Samyoul Scenario 1 only works correctly when devices are paired 1-2-1, but if paired 1 way then old sticker packs may get overridden on receiving device.
Scenario 2: also weird things happen.

when developed I had only live synchronization in mind. These scenarios have to be investigated.

**Pull request merged**
